File: multipeer/todo/ios/todo/Views/PeerListView.swift
//
//  PeerListView.swift
//  todo
//
//  Copyright (c) 2025 Couchbase, Inc All rights reserved.
//
//  Licensed under the Apache License, Version 2.0 (the "License");
//  you may not use this file except in compliance with the License.
//  You may obtain a copy of the License at
//
//  http://www.apache.org/licenses/LICENSE-2.0
//
//  Unless required by applicable law or agreed to in writing, software
//  distributed under the License is distributed on an "AS IS" BASIS,
//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
//  See the License for the specific language governing permissions and
//  limitations under the License.
//

import SwiftUI

struct PeerListView: View {
    @ObservedObject var viewModel: TodoViewModel

    var body: some View {
        NavigationStack {
            VStack {
                List(viewModel.peers) { peer in
                    HStack {
                        VStack(alignment: .leading) {
                            Text(peer.id.short())
                                .font(.headline)
                            Text(peer.status)
                                .font(.subheadline)
                                .foregroundColor(.gray)
                        }
                        Spacer()
                        Circle()
                            .fill(peer.online ? Color.green : Color.red)
                            .frame(width: 12, height: 12)
                    }
                    .padding(.vertical, 4)
                }
            }
            .navigationTitle("Peers")
            .toolbar {
                ToolbarItem(placement: .principal) {
                    VStack {
                        Text("\(viewModel.myPeerID.short())")
                            .font(.subheadline)
                            .foregroundColor(.gray)
                    }
                }
            }
        }
        .onAppear {
            viewModel.refreshPeers()
        }
    }
}

#Preview {
    PeerListView(viewModel: TodoViewModel())
}

File: multipeer/todo/android/app/src/main/java/com/couchbase/multipeer/todo/view/PeerListView.kt
//
//  Copyright (c) 2025 Couchbase, Inc All rights reserved.
//
//  Licensed under the Apache License, Version 2.0 (the "License");
//  you may not use this file except in compliance with the License.
//  You may obtain a copy of the License at
//
//  http://www.apache.org/licenses/LICENSE-2.0
//
//  Unless required by applicable law or agreed to in writing, software
//  distributed under the License is distributed on an "AS IS" BASIS,
//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
//  See the License for the specific language governing permissions and
//  limitations under the License.
//

package com.couchbase.multipeer.todo.view

import androidx.compose.foundation.background
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.lazy.LazyColumn
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.material3.ExperimentalMaterial3Api
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.ModalBottomSheet
import androidx.compose.material3.Text
import androidx.compose.material3.rememberModalBottomSheetState
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.collectAsState
import androidx.compose.runtime.getValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.platform.LocalConfiguration
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import com.couchbase.multipeer.todo.viewmodel.TodoViewModel
import com.couchbase.multipeer.todo.data.short
import com.couchbase.multipeer.todo.ui.theme.GreenOnline
import com.couchbase.multipeer.todo.ui.theme.RedOffline

@OptIn(ExperimentalMaterial3Api::class)
@Composable
fun PeerListView(viewModel: TodoViewModel, onDismiss: () -> Unit) {
    val peers by viewModel.peers.collectAsState()
    val sheetState = rememberModalBottomSheetState(skipPartiallyExpanded = true)

    val configuration = LocalConfiguration.current
    val targetHeight = configuration.screenHeightDp.dp * 0.9f

    LaunchedEffect(Unit) {
        viewModel.refreshPeers()
    }

    ModalBottomSheet(
        onDismissRequest = onDismiss,
        sheetState = sheetState
    ) {
        Column(modifier = Modifier
            .fillMaxWidth()
            .height(targetHeight)
            .padding(16.dp)) {
            Column(
                horizontalAlignment = Alignment.CenterHorizontally,
                modifier = Modifier.fillMaxWidth()
            ) {
                Text(
                    text = viewModel.myPeerID.short,
                    style = MaterialTheme.typography.bodySmall,
                    color = Color.Gray
                )
                Text(
                    text = "Peers",
                    style = MaterialTheme.typography.titleLarge
                )
            }

            Spacer(modifier = Modifier.height(16.dp))

            LazyColumn {
                items(peers.size) { index ->
                    val peer = peers[index]
                    Row(
                        modifier = Modifier
                            .fillMaxWidth()
                            .padding(vertical = 8.dp),
                        verticalAlignment = Alignment.CenterVertically
                    ) {
                        Column(modifier = Modifier.weight(1f)) {
                            Text(
                                peer.id.short,
                                style = MaterialTheme.typography.titleMedium.copy(fontSize = 18.sp)
                            )
                            Text(
                                peer.status,
                                style = MaterialTheme.typography.bodySmall.copy(fontSize = 16.sp),
                                color = Color.Gray
                            )
                        }
                        Box(
                            modifier = Modifier
                                .size(12.dp)
                                .clip(CircleShape)
                                .background(if (peer.online) GreenOnline else RedOffline)
                        )
                    }
                }
            }
        }
    }
}